### ISSUE TITLE
Generalizations and extensions

### ISSUE THREAD

**Issue by aplavin** [oschulz/PropertyFunctions.jl#4]
I think the approach taken in this package is a great idea! Let me start some discussion about a somewhat wider picture.

This was a piece I was missing for quite some time. There has already been a way to define "introspectable" property-getter functions that can in principle be optimally applied to columnar arrays – `Accessors.PropertyLens(:propname)`. With them, stuff like `map((@optic _.a.b + 1), structarray)` could be made performant in a trivial way. However, this is not that useful really, so such `map` methods weren't even implemented.

`PropertyFunctions.jl` took it to an actually useful level, making it so that applying more complex functions (like `@pf $a + $c^2`) to structarrays only accesses relevant data. I recently noticed this package, and took the liberty to generalize/adapt its features – for now they are in `AccessorsExtra`.
- Adaptation: mainly, the syntax is same as used in `Accessors` optics, with `_` instead of `$`: `@o _.a.b + 1 + _.c`. So it's just one macro and one syntax to define all these "introspectable functions".
- Generalization: nested property accesses are supported, see these examples for StructArrays: https://gitlab.com/aplavin/AccessorsExtra.jl/-/blob/master/test/propertyfunction.jl#L78.
- Also, with a few small package extensions, they work with both `StructArrays` and `DictArrays` (both arbitrarily nested), with `mapview` (and automatically `groupview`), + some more, https://gitlab.com/aplavin/AccessorsExtra.jl/-/blob/master/test/propertyfunction.jl.

These features are quite new and not mature yet: eg `filter` should also be defined, similar to how you do it here; `FlexiJoins` support would also be nice. But I quite like how it plays out already.
Please let me know what you think about these and other potential generalizations, or about relevant design decisions! Potentially, it would be nice to have `PropertyFunction` itself defined here with all relevant functions/`@pf`/overloads and with no dependencies, while `AccessorsExtra` would just handle its construction in the `@optic` macro in a consistent way.

**Comment by oschulz** [comment]
Hi @aplavin , thanks for your kind words!

PropertyFunctions actually grew out of data-processing use cases I have that involve tables with columns which are expensive to access, so I wanted to make it easy for users to broadcast simple functions that touch only what's necessary (I actually don't have it so bad, ask @moelf for data that can have thousands of columns, only a small subset of which may need to be accessed at a time). The fun thing is that it was quite easy to make PropertyFunctions work on both GPU and GPU, at least with `StructArray`s. The reason for using "$a" instead of something like "_.a" is was to make the expression as compact as possible for the user (`$a + $c^2` may perhaps be more readable than ` _.a + _.c^2`, but the Accessor-syntax is of course more flexible in general).

I didn't have any major ambitions beyond the current state originally, but I like your ideas very much. And nested property access is definitely great to have, as well as having more advanced operations (grouping, etc.). I'd be very open to extend/generalize in conjunction with the "Accessor-verse".  Beyond adding this kind of functionality to Accessors(-Extra), do you think we could provide some integration? For example, could we make Accessors use/convert `PropertyFunction`s generated with `@pf`, e.g. via a package extension here or there?

**Comment by Moelf** [comment]
I forgot about this package sorry Oliver, but I like the discussion I'm seeing here, specifically two things I resonate with a lot:
- only accesses relevant data
- in principle be optimally applied to columnar arrays

the first is important because our columns can be disk-backed and loading every column is silly when user may only need 10% of them. The second I recently started looking into because there's interests in having query-like way doing things as oppose to loops-tyle, and in the process, I realized `Query.jl` is "lazy" but you're limited to a handful of operations and also the erganomics is not the best.

I have a concrete example for people who are interested:
https://github.com/Moelf/UnROOT_RDataFrame_MiniBenchmark/blob/8492a186d16504b5a510ad1ef007c68a4f11e004/composite_benchmarks/UnROOT_loop.jl#L13

and the corresponding `Query.jl` style looks like this:
```julia
julia> function query_main(t)
           @from evt in t begin
               @where evt.nMuon == 4
               @let pts = evt.Muon_pt
               @let etas = evt.Muon_eta
               @where all(pts .> 5) && all(abs.(etas) .< 2.4)
               @let phis = evt.Muon_phi
               @let masses = evt.Muon_mass
               @let charges = evt.Muon_charge
               @where sum(charges) == 0
               @let Z_idx = reco_zz_to_4l(pts, etas, phis, masses, charges)
               @where filter_z_dr(Z_idx, etas, phis)
               @let  Z_mass = compute_z_masses_4l(Z_idx, pts, etas, phis, masses)
               @where filter_z_candidates(Z_mass)
               @let h_mass = compute_higgs_mass_4l(Z_idx, pts, etas, phis, masses)

               @select h_mass
               @collect
           end
       end
```

and the run time is similar. I'm sure we can try to use PF.jl too, but my point is I'm generally intereted in this (design) space of optimal access & optimal execution for basic-ish user-defined data crunching
```

**Comment by oschulz** [comment]
> I forgot about this package sorry Oliver, but I like the discussion I'm seeing here, specifically two things I resonate with a lot

My fault, Jerry, I forgot to talk to you about it at the JuliaHEP meeting. I always wanted to try out PropertyFunctions with UnROOT, just hadn't gotten around to it yet ... I might just work out of the box, and if not it probably won't require much to make it work.

**Comment by oschulz** [comment]
> the user-written event loops probably look like:

Oh, sure - I thought more in the context of broadcast syntax, like

```julia
@pf($col1 + $col2).(tree)
```

or

```julia
filterby(view, @pf $col1 + $col2 < $col10)(tree)
```

**Comment by Moelf** [comment]
right, these are nice but doesn't fit into the workflow -- would user really pick up `@pf` and drop it later when the analysis needs evolve? and learning `@pf` as oppose to just stick with Base Julia also adds mental burden to user.

These are the design-space problem I'm struggling with I guess, I don't want user to re-write using different Julia package/stype when going from "simple inspection" to "full analysis" and I don't see `@pf` being part of the recipe right now

**Comment by Moelf** [comment]
for that I can see user opt for the vanilla Julia stuff just because it doesn't require learning a new package:

```julia
@pf($col1 + $col2).(tree)
# 
(x -> x.col1 + x.col2).(tree)
```

```julia
filterby(view, @pf $col1 + $col2 < $col10)(tree)
# 
filter(tree) do evt
    evt.col1 + evt.col2 < evt.col10
end
```

**Comment by oschulz** [comment]
Yes, but `(x -> ...).(tree)` will unfortunately (except in very simple where the compiler can optimize it away) load all rows. ;-)

**Comment by oschulz** [comment]
> Really not looking forward to add StructArrays and Tables as dependencies there, wish they weren't here either.

Me too! :-) I didn't find a way to get what I wanted (GPU support and output of StructArrays even when broadcasting over something like DataFrame columns) without a directly dependency, but it certainly makes PropertyFunctions heavier than I'd like.

**Comment by oschulz** [comment]
> Hmm, what's wrong with simply defining Base.filter(::PropertyFunction, X)

I thought about that, but I wanted to also allow the curried form `filterby(pf)(x)` to make operations easily chainable. I thought maybe it would be better not to bend the `Base.filter` API that far, especially since it might then be confusing why `filter(f)` works with some functions but not others.

**Comment by oschulz** [comment]
> I can understand the confusion :) but none of this functionality is in Accessors proper. 

Oh, right! :-) 

> Conversion can be done in both directions

Great - I'd be happy to make AccessorsExtra a weak-dep of PropertyFunctions.

> in this package, pf.sel_prop_func takes a Tuple of extracted values, while in AccessorsExtra pf.func takes an object (NamedTuple) with requested properties. Makes nested properties handling easier.

Yes, I like that approach. One reason why I went for the "flat tuples" was that I hope it might introduce less overhead during automatic differentiation (esp. with Zygote), because things are "unpacked" only once, so no intermediate tangents of NamedTuples with only a few "active fields". But I would like to support `$a.b.c` in the future ... 

> so it's not really a lens. For example, PF cannot be used to modify a part of an object:

Oh, of course!

**Comment by oschulz** [comment]
> filter(f)(X) works with all functions automatically, there's this definition in Base:

Yes, unfortunately not on older Julia versions though, for example not on v1.6-LTS, and Compat doesn't backport it either. :-( Not sure when it was added, might have been after I designed PropertyFunctions. Oh and `filterby` also supports "view-mode", which `filter` doesn't.

**Comment by oschulz** [comment]
> Any specific test you have in mind to ensure PF works on GPU?

It "just works" as `PropertyFunction` is written, because broadcasting a function over several flat GPU array just works. And StructArrays does a little `BroadcastStyle` thing to be GPU-broadcast compatible.

> > and output of StructArrays even when broadcasting over something like DataFrame columns
>This is clearly not possible without a StructArrays dep. But is such behavior needed/expected?

I like it because I think the output should be columnar for both `@pf((a = ..., b = ...)).(some_table)` and `@pf((a = ..., b = ...)).(simple_vector)`. It makes things consistent, and StructArrays is the best solution we currently have, I think. When broadcasting something that produces `NamedTuple`s over a plain vector of numbers, there's nothing in the input that we can dispatch on to decide what output type we should use. So I just  go for `StructArray` in all cases where the output of the function is a `NamedTuple`. :-)on

StructArrays is a bit like StaticArrays I think - the best representation we have for each use case, and therefore hard to avoid sometimes. Fortunately StructArrays has become more lightweight (though not super-lightweight) over time, in contrast to StaticArrays. :-)

**Comment by oschulz** [comment]
> for now it exists in Skipper.jl, called filterview. Generally useful when working with collections, but doesn't seem that good of a fit to Base IMO...

Yes, I don't think we should stretch the API of `Base.filter` that far.

**Comment by oschulz** [comment]
> Then, arrays result in arrays, StructArrays in StructArrays, etc

Yes, that work too :-)

**Comment by Moelf** [comment]
>Interesting, do you use type-unstable containers for that? If so, what functions do you use to work with them performantly?
I don't remember seing tables with more than a few hundred columns myself.

we use type-stable container, basically we have our own `TypedTables.Table`, and we roll out own `StructArray.LazyRow` to delay access inside the loop until user call `getproperty(row, ...)`. Luckily, building the table for 1500 columns [only takes ~1s in 1.10](https://indico.cern.ch/event/1292759/contributions/5618591/attachments/2747123/4780773/UnROOT_Past_2023_Future.pdf#page=37).

>IMO seems a bit strange to use a heavy macro-based package like Query.jl, but be hesitant to adopt a small self-contained @pf or @o macro.

user may want to learn some ecosystem similar to Query.jl with the hole they can do everything there, like SQL. While `@pf` `@o` do not add enough value, but are still macro (not composable).

**Comment by oschulz** [comment]
> while @pf @o do not add enough value, but are still macro (not composable).

I would say it depends very much on the use case. :-)

**Comment by Moelf** [comment]
yeah I'm just speaking from the very narrow, LHC-centric analysis use case

**Comment by Moelf** [comment]
https://github.com/Moelf/ADLBenchmark.jl

it seems to me either I just write for-loop or I'm doing column-at-a-time quick study. For former I can't use PF.jl, for latter PF.jl is not so much better than just vanilla Julia functions

**Comment by Moelf** [comment]
I mean by no means this is a problem or deficiency of PF.jl -- this is just generally a hard problem.

There doesn't exist a design pattern or technology that can organically combine loopy code and columnar kernel anywhere. One can try to start from either side, e.g. Python people they start from the columnar side, and then they run into hard-to-express computation, they may then write a Numba/Jax kernel, but that kernel may be as complex as your loop-body.

In Julia we naturally start from the loopy code, but then one just never bother with PF.jl

**Comment by Moelf** [comment]
it's good to have examples, so this is what I mean, users start out looking at one histogram. Eventually it will evolve into looking at O(10)xO(100) histograms, the first O(10) are different "signal regions", the second O(100) is how many different systematics you deal with:
https://github.com/Moelf/WVZAnalysis.jl/blob/master/src/mainlooper.jl#L251-L297

for maximal efficiency, we want to loop the data only once of course, so you need to have `push!()` in the middle of things, which breaks the `DataPipes` way of thinking about things.

Anyway, this is not a problem of PF.jl, it's a hard engineering/design issue that nobody has good answer for, here are three languages (frameworks) doing the same semi-realistic analysis trying to figure out what are some good patterns:
- https://github.com/iris-hep/analysis-grand-challenge (python)
- https://github.com/Moelf/LHC_AGC.jl (julia)
- https://github.com/andriiknu/RDF (C++ ROOT)

**Comment by Moelf** [comment]
>Ehm, why does it "break" anything? Iterators.map/filter are specifically designed to make single-pass processing, and I'm pretty sure they can be made to work efficienly with PF.

my understanding is DataPipes.jl assumes push-down all the way, in the same way. You're looking at branching `push!()` and it's not ergonomic to accumulate 1000 numbers every iteration and push everything in the end, put it another way, our event loops are not (in the general sense) SIMD friendly, you accumulate to different subset of histograms based on the content of each iteration. 

- - -
>Sounds like a great case for Julia and packages,

that's the goal

**Comment by oschulz** [comment]
> In Julia we naturally start from the loopy code, but then one just never bother with PF.jl

I wouldn't subscribe to that as a given. :-) Sure, in Julia we can and do write fast loops, but that doesn't mean we have to write loops all the time, e.g. when broadcasts and similar will do the trick. ;-)

**Comment by Moelf** [comment]
For example, I think these both suck, but at least in the loopy version, it's easy for a student to go in and change something about this algorithm

```python
        events["Electron", "pdgId"] = -11 * events.Electron.charge
        events["Muon", "pdgId"] = -13 * events.Muon.charge
        events["leptons"] = ak.concatenate([events.Electron, events.Muon], axis=1,)
        events = events[ak.num(events.leptons) >= 3]

        pair = ak.argcombinations(events.leptons, 2, fields=["l1", "l2"])
        pair = pair[(events.leptons[pair.l1].pdgId == -events.leptons[pair.l2].pdgId)]
        with np.errstate(invalid="ignore"):
            pair = pair[
                ak.singletons(
                    ak.argmin(
                        abs(
                            (events.leptons[pair.l1] + events.leptons[pair.l2]).mass
                            - 91.2
                        ),
                        axis=1,
                    )
                )
            ]
        events = events[ak.num(pair) > 0]
        pair = pair[ak.num(pair) > 0][:, 0]

        l3 = ak.local_index(events.leptons)
        l3 = l3[(l3 != pair.l1) & (l3 != pair.l2)]
        l3 = l3[ak.argmax(events.leptons[l3].pt, axis=1, keepdims=True)]
        l3 = events.leptons[l3][:, 0]

        mt = np.sqrt(2 * l3.pt * events.MET.pt * (1 - np.cos(events.MET.delta_phi(l3))))
```

```julia
               leptons = Vcat(
			LorentzVectorCyl.(evt.Electron_pt, evt.Electron_eta, evt.Electron_phi, evt.Electron_mass),
			LorentzVectorCyl.(evt.Muon_pt, evt.Muon_eta, evt.Muon_phi, evt.Muon_mass)
		)
		nLep = length(leptons)
		nLep < 3 && continue
		lep_pids = Vcat(evt.Electron_charge .* 11, evt.Muon_charge .* 13)

		sel_combo = [1, 1]
		best_diff = Inf
		
		for i in 1:nLep
			i_pid = lep_pids[i]
			i_lep = leptons[i]
			for j in 1+i:nLep
				i_pid != -lep_pids[j] && continue
				temp_mass = fast_mass(i_lep, leptons[j])
				temp_diff = abs(temp_mass - Z_m)
				if temp_diff < best_diff
					best_diff = temp_diff
					sel_combo .= i, j
				end
			end
		end
		
		isinf(best_diff) && continue # no Z candidate found
		
		i_lep3 = argmax(eachindex(leptons)) do i
			i ∈ sel_combo && return -Inf32
			return pt(leptons[i])
		end
		l3 = leptons[i_lep3]
		
		mt = sqrt(2 * l3.pt * evt.MET_pt * (1.0 - cos(evt.MET_phi - l3.phi)))
```

**Comment by Moelf** [comment]
>These are quite nice to work in Julia actually! When your data is "small" (so that all operations can be materalizing), all functionality to do that already exists. This is a major area where I feel much better than using Python. And no for-loops either :)

sorry, maybe you're not understanding the big picture, I'm saying that's 1 column, and your Table has 1000 columns like that, and your "analysis" is to loop over all these columns once (because you want to minimize things), while doing row-level computation

**Comment by Moelf** [comment]
>Higher-level code is so much easier to understand (compared to for loops) in other fields of data analysis

I'm saying this sentence is super meaningful, by definition high-level code is easier to understand, but if you actually try to do a "HEP data analysis", you will find you can't express your problem in the nice paradigms we're familiar with.

The "data analysis" is fundamentally "row-level computation", and your "row" is hundreds of vectors of structs

**Comment by Moelf** [comment]
>There is probably no readily implemented stuff for lazy/singlepass operations like these, but there are no fundamental limitations.

there is, Query.jl does everything in a single pass whenever possible, but it's quite limited for many reasons as we already knew. So again, unless we have Coffea or RDataFrame re-implemented in Julia, it's just not practical to do anything other than for-loop -- given you know you're doing a full, serious HEP analysis at the end of the day

**Comment by Moelf** [comment]
sorry, I'm not being helpful nor contributing to the original topic constructively, we should move on.

**Comment by oschulz** [comment]
I think this have gone a bit beyond the scope of this issue. @Moelf does indeed have use cases where it's kind of hard to achieve the required "loop inversion" (doing many different things for the same data with different output, and doing all of that for each bit of data) automatically. But on the other hand, in many use cases (I'd say also quite a few in HEP, depends on the level of the analysis), columnar/broadcast-style will do fine. The nice thing in Julia is that we can do whatever fits the use case best, without worring too much. :-)

**Comment by oschulz** [comment]
The latest version of PropertyFunctions comes with Accessors integration, nested property access and some other improvements - I hope you like it, @aplavin .

The new version also has GPU/Reactant support.

Closing this issue, though maybe we'll want to create some specific follow-up issues.
